src/qmapshack/realtime/ais/CRtAisInfo.cpp

```diff
@@ -0,0 +1,591 @@
+
+    emit sigChanged();
+
+    //qWarning() << "B, MMSI:" << ship->mmsi << ", Lat: " << ship->latitude << ", Lon: " << ship->longitude;
+}
+
+void CRtAisInfo::ais21(const QByteArray& data)
+{
+    CRtAis* _source = dynamic_cast<CRtAis*>(source.data());
+
+    ais_aid_to_navigation_t ais;
+    ais.type = data[0];
+    ais.repeat = getInt(data, 6, 2);
+    ais.mmsi = getInt(data, 8, 30);
+    ais.aidType = getInt(data, 38, 5);
+    getString(data, ais.name, 43, 120);
+    ais.accuracy = getSignedInt(data, 163, 1);
+    ais.lon = getSignedInt(data, 164, 28);
+    ais.lat = getSignedInt(data, 192, 27);
+    ais.dimToBow = getInt(data, 219, 9);
+    ais.dimToStern = getInt(data, 228, 9);
+    ais.dimToPort = getInt(data, 237, 6);
+    ais.dimToStarboard = getInt(data, 243, 6);
+
+    bool ok = true;
+    CRtAis::ship_t* ship = _source->getShipByKey(QString::number(ais.mmsi) , ok);
+    ship->longitude = ais.lon / 600000.0;
+    ship->latitude = ais.lat / 600000.0;
+    ship->pos = QPointF(ship->longitude, ship->latitude);
+    ship->timePosition = QDateTime::currentSecsSinceEpoch();
+    ship->name = ais.name;
+    ship->aid = true;
+    ship->heading = -1;
+    ship->velocity = -1;
+
+    lastTimestamp = QDateTime::currentDateTime();
+
+    emit sigChanged();
+}
+
+void CRtAisInfo::ais24(const QByteArray& data)
+{
+    CRtAis* _source = dynamic_cast<CRtAis*>(source.data());
+
+
+    ais_static_and_voyage_t ais;
+    ais.type = data[0];
+    ais.repeat = getInt(data, 6, 2);
+    ais.mmsi = getInt(data, 8, 30);
+
+    uint8_t part = getInt(data, 38, 2);
+    if(part == 0)
+    {
+        getString(data, ais.shipName, 40, 120);
+
+        bool ok = false;
+        CRtAis::ship_t* ship = _source->getShipByKey(QString::number(ais.mmsi) , ok);
+        if(ok)
+        {
+            ship->name = ais.shipName;
+
+            lastTimestamp = QDateTime::currentDateTime();
+
+            emit sigChanged();
+
+            //qWarning() << "BA, MMSI: " << ais.mmsi << ", name: " << ais.shipName;
+        }
+    }
+    else if(part == 1)
+    {
+        ais.shipType = getInt(data, 40, 8);
+        getString(data, ais.callsign, 90, 42);
+        ais.dimToBow = getInt(data, 132, 9);
+        ais.dimToStern = getInt(data, 141, 9);
+        ais.dimToPort = getInt(data, 150, 6);
+        ais.dimToStarboard = getInt(data, 156, 6);
+
+        bool ok = false;
+        CRtAis::ship_t* ship = _source->getShipByKey(QString::number(ais.mmsi) , ok);
+        if(ok)
+        {
+            ship->callsign = ais.callsign;
+
+            lastTimestamp = QDateTime::currentDateTime();
+
+            emit sigChanged();
+
+            //qWarning() << "BB, MMSI: " << ais.mmsi << ", callsign: " << ais.shipName;
+        }
+    }
+}
+
+uint32_t CRtAisInfo::getInt(const QByteArray& data, int start, int count)
+{
+    int mask = 0x3f;
+
+    int end = start + count;
+
+    int from = start / 6;
+    int to = end / 6;
+
+    int from_mask_ignore_bits = start % 6;
```

Comment:
I guess this code is copy-n-paste from another project. Do we have the copy right for that? Do we need to mention the original author?

Well and... nameing conventions.  We do camelCase for variables.

Reply:
I am the original author, but yes, copy paste from a java project I made probably 10+ years ago that decodes AIS :)